src/java/com/puppetlabs/jruby_utils/pool/JRubyPool.java

```diff
@@ -491,8 +491,7 @@ private void signalIfLockCanProceed() {
         // is active at a time - a caller of lock() that has just acquired
         // the pool lock but is waiting for all registered elements to be
         // returned to the queue.
-        if (registeredElements.size() == liveQueue.size() ||
-                pill != null) {
+        if (this.maxSize == liveQueue.size() || pill != null) {
```

Comment:
Thanks for the explanation @camlow325! I think I was missing the bit that we were originally using `registeredElements` rather than `maxSize` in the past because we hoped to do dynamic sizing at some point - if that was our only requirement for having that then this change all makes sense to me (and I'm less scared about it :D). 👍